Improve security failure handling in gRPC

michalvavrik opened the pull request:
- fixes the issue reported here https://quarkusio.zulipchat.com/#narrow/channel/187030-users/topic/Unhandled.20exception.20in.20router.20quarkus.20security.20and.20grpc.2E - which is main router not having an error handler means that when proactive authentication fails it resulted in unhandled exception (500 instead of 4xx)
- and the test I added uncovered couple of more things in gRPC security exception handling, this PR:
  - makes sure that security  and authentication exceptions are not added as description to status unless in DEV / TEST. it's not good idea to give away details about failure like exception message in production and https://grpc.io/docs/guides/error/#protocol-errors says that description is optional; users have a choice to define their own exception handler or even authentication exception handler and add them
  - when using separate gRPC server and no gRPC security mechanism provides identity, an anonymous identity is set as we don't work with null identity in Quarkus Security
  - when gRPC is combined with other REST stack like Quarkus REST and lazy authentication failed, I have mentioned that there was a race in the exception handling as authentication failure handler needs to be removed, otherwise Ver.tx failure handler chain takes over; when gRPC starts processing, it can handle authentication failures on it's own (in logs there was _response already written_)
  - fixes response status for authentication and security failures that happens when authorization using standard security annotations triggers authentication it ended in the `io.quarkus.grpc.runtime.supports.exc.DefaultExceptionHandlerProvider` and the `io.quarkus.grpc.auth.AuthExceptionHandlerProvider` was never invoked for that failure; I corrected the response status to UNAUTHENTICATION/PERMISSION_DENIED, before this change it was UNKNOWN for authorization and authentication exceptions as well
- making sure that authorization handler is not skipped even though we don't support that according to the docs, it was either that or dropping the handler, context here https://github.com/quarkusio/quarkus/pull/45861#discussion_r1930317867

michalvavrik committed:
fix(grpc,security): Improve security error handling

cescoffier reviewed (approved):
Looks good!

@alesj WDYT?

alesj commented on extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/GrpcServerRecorder.java:
```diff
@@ -198,12 +222,13 @@ private void buildGrpcServer(Vertx vertx, GrpcServerConfiguration configuration,
         if (securityHandlers != null) {
             for (Map.Entry<Integer, Handler<RoutingContext>> e : securityHandlers.entrySet()) {
                 Handler<RoutingContext> handler = e.getValue();
+                boolean isAuthenticationHandler = e.getKey() == -200;
```
What's this "-200" ?

michalvavrik replied:
It is 200 * -1, see:

- https://github.com/quarkusio/quarkus/blob/a6413e331c3b072c9e624d93189772341d2547ad/extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/FilterBuildItem.java#L16
- https://github.com/quarkusio/quarkus/blob/a6413e331c3b072c9e624d93189772341d2547ad/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java#L430

I can imagine it feels wrong, but if I introduce runtime constant, this place will be only consumer anyway, so it won't give anyone hint that they need to update it in this class. FWIW changing priority would be super breaking, but I can make it constant if you like.

michalvavrik replied:
Or give it order and know that the first handler is the authenticator. I think -200 is acceptable, but if you will different, just let me know and I'll rewrite it to avoid it entirely. np.

gsmet commented:
@michalvavrik I tried to backport this one and it ended up failing.

You can see the failures in this hidden comment: https://github.com/quarkusio/quarkus/pull/46342#issuecomment-2667158037 .

michalvavrik commented:
It also fails on main, CI of this PR probably didn't run it and even though I added this test in past I forgot about it. Sorry, is 21 IT gRPC modules and I also added some gRPC tests into different modules like `oidc-wiremock` to test some specific behavior. I should have run all related IT modules and checker properly, especially as it is called `grpc-mutual-auth`. I'll be more careful.

Here is what happens:

- previously response status was UNKNOWN and stacktrace contained Forbidden, no response status is PERMISSION_DENIED and stack contains 

I checked and `io.quarkus.grpc.runtime.supports.exc.DefaultExceptionHandlerProvider` transform excetion into `StatusRuntimeException: PERMISSION_DENIED` with status description being set to `io.grpc.StatusRuntimeException: PERMISSION_DENIED` stacktrace. Thrown `io.quarkus.security.ForbiddenException` has empty exception message, which is why there is no status description.

I think this behavior is better because now behavior is according to the https://grpc.io/docs/guides/status-codes/ (you have correct response status). The stacktrace of the `io.quarkus.security.ForbiddenException` is not very useful unless you know Quarkus implementation, which is why status description only contains exception message when present (that didn't change, previously it was always set to the exception message).

My suggestion is that I fix the test. The PR reviewers can consider situation and request something else, for example we could start adding exception stacktrace into status description as well. Here it is: https://github.com/quarkusio/quarkus/pull/46350
